web-common/src/components/button-toggle-group/SubButton.svelte

```diff
@@ -0,0 +1,72 @@
+<script lang="ts">
+  import { getContext } from "svelte";
+  import {
+    ButtonGroupContext,
+    buttonGroupContext,
+  } from "./ButtonToggleGroup.svelte";
+
+  import Tooltip from "../tooltip/Tooltip.svelte";
+  import TooltipContent from "../tooltip/TooltipContent.svelte";
+
+  export let key: number | string;
+  export let tootips: {
+    selected?: string;
+    unselected?: string;
+    disabled?: string;
+  } = undefined;
+
+  const {
+    registerSubButton,
+    toggleSubButton,
+    selectedKey,
+    firstKey,
+    lastKey,
+    disabledKeys,
+  }: ButtonGroupContext = getContext(buttonGroupContext);
+
+  registerSubButton(key);
+
+  $: isDisabled = disabledKeys.includes(key);
+  $: isSelected = $selectedKey === key;
+
+  const baseStyles = `shrink flex flex-row items-center px-1 py-1
+  transition-transform duration-100`;
+
+  $: textStyle = isDisabled
+    ? "text-gray-400"
+    : "text-gray-700 hover:text-gray-900 ";
+
+  $: bgStyle = isDisabled
+    ? "bg-white"
+    : isSelected
+    ? "bg-gray-100 hover:bg-gray-200 "
+    : "bg-white hover:bg-gray-50 ";
+
+  // This is needed to make sure that the left and right most child
+  // elements don't break out of the border drawn by the parent element
+  $: isFirst = key === $firstKey;
+  $: isLast = key === $lastKey;
+  $: roundings = `${isFirst ? "rounded-l" : ""} ${isLast ? "rounded-r" : ""} `;
+
+  $: finalStyles = `${baseStyles} ${roundings} ${textStyle} ${bgStyle}`;
+
+  $: tooltipText = isDisabled
+    ? tootips?.disabled
+    : isSelected
+    ? tootips?.selected
+    : tootips?.unselected;
+</script>
```

Comment:
This component never calls the context's "onDestroy()" method to deregister itself. Was that on purpose?
Doesn't it need a lifecycle hook here to unregister itself from the ButtonGroup if it gets removed?

Reply:
Good catch. The source I was going off of (which was recommended by none other than Rich Harris, and I suspect written by him!) also forgot to call onDestroy in the child elts, and since I was following a template from that source, I must admit I didn't look too deeply at everything. Very nice to have a second pair of eyes! :-)

Reply:
hmm TIL... never mind it turns out that what they were doing was a bit more subtle than I thought. Also it turns out that Svelte is once again a bit mind-twisting for me.

Checked the docs at https://svelte.dev/tutorial/ondestroy -- the key line is "While it's important to call lifecycle functions during the component's initialisation, it doesn't matter _where_ you call them from." -- I think the `onDestroy` method is called during the initialization of the the subbutton component (since `registerSubbutton` is called during initialization) even though the code is located in the parent component.

There are some patterns in Svelte that I really do not like...

Reply:
Wow, that is kinda mind blowing. I did not catch that `onDestroy` was within `registerSubbutton`